api/handlers.go

```diff
@@ -0,0 +1,156 @@
+package server
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"strconv"
+	"strings"
+)
+
+// GoCertRouter takes in an environment struct, passes it along to any handlers that will need
+// access to it, then builds and returns it for a server to consume
+func GoCertRouter(env *Environment) http.Handler {
+	router := http.NewServeMux()
+	router.HandleFunc("GET /certificate_requests", GetCertificateRequests(env))
+	router.HandleFunc("POST /certificate_requests", PostCertificateRequest(env))
+	router.HandleFunc("GET /certificate_requests/{id}", GetCertificateRequest(env))
+	router.HandleFunc("DELETE /certificate_requests/{id}", DeleteCertificateRequest(env))
+	router.HandleFunc("POST /certificate_requests/{id}/certificate", PostCertificate(env))
+
+	v1 := http.NewServeMux()
+	v1.HandleFunc("GET /status", HealthCheck)
+	v1.Handle("/api/v1/", http.StripPrefix("/api/v1", router))
+
+	return logging(v1)
+}
+
+// the health check endpoint simply returns a 200
+func HealthCheck(w http.ResponseWriter, r *http.Request) {
+	w.WriteHeader(200) //nolint:errcheck
+}
+
+// GetCertificateRequests returns all of the Certificate Requests
+func GetCertificateRequests(env *Environment) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		certs, err := env.DB.RetrieveAll()
+		if err != nil {
+			logError(err.Error(), 500, w)
+			return
+		}
+		body, err := json.Marshal(certs)
+		if err != nil {
+			logError(err.Error(), 500, w)
+			return
+		}
+		if _, err := w.Write(body); err != nil {
+			logError(err.Error(), 500, w)
+		}
+	}
+}
+
+// PostCertificateRequest creates a new Certificate Request, and returns the id of the created row
+func PostCertificateRequest(env *Environment) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		csr, err := io.ReadAll(r.Body)
+		if err != nil {
+			logError(err.Error(), 500, w)
+			return
+		}
+		id, err := env.DB.Create(string(csr))
+		if err != nil {
+			if strings.Contains(err.Error(), "UNIQUE constraint failed") {
+				logError("given csr already recorded", 400, w)
+				return
+			} else {
+				logError(err.Error(), 400, w)
+				return
+			}
+		}
+		w.WriteHeader(201)
+		if _, err := w.Write([]byte(strconv.FormatInt(id, 10))); err != nil {
+			logError(err.Error(), 500, w)
+		}
+	}
+}
+
+// GetCertificateRequests receives an id as a path parameter, and
+// returns the corresponding Certificate Request
+func GetCertificateRequest(env *Environment) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		id := r.PathValue("id")
+		cert, err := env.DB.Retrieve(id)
+		if err != nil {
+			if err.Error() == "sql: no rows in result set" {
+				logError(err.Error(), 400, w)
```

Comment:
Shoulnd't we return 404 here?

Reply:
I think 404 could be confusing, since it means that the path isn't available. Users may assume that the endpoint is wrong rather than the request parameters. Maybe a 204 since the request was valid?